src/main/kotlin/com/sourcegraph/cody/history/HistoryService.kt

```diff
@@ -0,0 +1,55 @@
+package com.sourcegraph.cody.history
+
+import com.intellij.openapi.components.*
+import com.sourcegraph.cody.agent.protocol.ChatMessage
+import com.sourcegraph.cody.agent.protocol.Speaker
+import com.sourcegraph.cody.history.state.ChatState
+import com.sourcegraph.cody.history.state.HistoryState
+import com.sourcegraph.cody.history.state.MessageState
+import java.time.LocalDateTime
+
+@State(name = "ChatHistory", storages = [Storage("cody_history.xml")])
+@Service(Service.Level.PROJECT)
+class HistoryService : SimplePersistentStateComponent<HistoryState>(HistoryState()) {
+
+  private val listeners = mutableListOf<(ChatState) -> Unit>()
+
+  fun listenOnUpdate(listener: (ChatState) -> Unit) {
+    synchronized(listeners) { listeners += listener }
+  }
+
+  fun update(internalId: String, chatMessages: List<ChatMessage>) {
+    val found = getChatOrCreate(internalId)
+    found.messages = chatMessages.map(::convertToMessageState).toMutableList()
+    found.setUpdatedTimeAt(LocalDateTime.now())
+    listeners.forEach { it(found) }
+  }
+
+  fun remove(internalId: String) {
+    state.chats.removeIf { it.internalId == internalId }
+  }
+
+  private fun convertToMessageState(chatMessage: ChatMessage): MessageState {
+    val message = MessageState()
+    message.text = chatMessage.text
+    message.speaker =
+        when (chatMessage.speaker) {
+          Speaker.HUMAN -> MessageState.SpeakerState.HUMAN
+          Speaker.ASSISTANT -> MessageState.SpeakerState.ASSISTANT
+        }
+    return message
+  }
+
+  private fun getChatOrCreate(internalId: String): ChatState {
```

Comment:
Can we use proper type instead of string?
I mean, I know it's a string now, but we can keep it UUID, and this way we will never accidentally mix it up with `sessionId`.